Apromore-Core-Components/Apromore-Manager/src/main/java/org/apromore/service/impl/WorkspaceServiceImpl.java

```diff
@@ -695,26 +697,125 @@ private void removeGroupLog(Group group, Log log) {
         }
     }
 
-    // TODO: whether should I put this logic here or in a separate service?
-//    private void createGroupUsermetadata(Group group, UserMetadata userMetadata, boolean hasRead, boolean hasWrite,
-//                                         boolean hasOwnership) {
-//        GroupUserMetadata groupUserMetadata = groupUserMetadataRepo.findByGroupAndUserMetadata(group, userMetadata);
-//        if (groupUserMetadata == null) {
-//            groupLog= new GroupLog(group, log, hasRead, hasWrite, hasOwnership);
-//            log.getGroupLogs().add(groupLog);
-//            //group.getGroupLogs().add(groupLog);
-//        } else {
-//            groupLog.setHasRead(hasRead);
-//            groupLog.setHasWrite(hasWrite);
-//            groupLog.setHasOwnership(hasOwnership);
-//        }
-//        groupLogRepo.save(groupLog);
-//    }
-//
-//    private void removeGroupUsermetadata(Group group, UserMetadata usermetadata) {
-//        GroupLog groupLog = groupLogRepo.findByGroupAndLog(group, log);
-//        if (groupLog != null) {
-//            groupLogRepo.delete(groupLog);
-//        }
-//    }
+    public Map<Group, AccessType> getLogACL(Integer logId) {
+
+        Map<Group, AccessType> groupAccessTypeMap = new HashMap<>();
+
+        for (GroupLog g : getGroupLogs(logId)) {
+            if (g.getHasRead() && g.getHasWrite() && g.getHasOwnership()) {
+                groupAccessTypeMap.put(g.getGroup(), AccessType.OWNER);
+            } else if (g.getHasRead() && g.getHasWrite()) {
+                groupAccessTypeMap.put(g.getGroup(), AccessType.EDITOR);
+            } else if (g.getHasRead()) {
+                groupAccessTypeMap.put(g.getGroup(), AccessType.VIEWER);
+            } else {
+                groupAccessTypeMap.put(g.getGroup(), AccessType.NONE);
+            }
+        }
+
+        return groupAccessTypeMap;
+    }
+
+    public Map<Group, AccessType> getProcessACL(Integer processId) {
+
+        Map<Group, AccessType> groupAccessTypeMap = new HashMap<>();
+
+        for (GroupProcess g : getGroupProcesses(processId)) {
+            if (g.getHasRead() && g.getHasWrite() && g.getHasOwnership()) {
```

Comment:
Repeated code

Reply:
Are there ways to reduce repeating code here other than abstract comparing and use ternary operator? Thanks

Reply:
This block of code: {if...then...else...} with hasRead, hasWrite, hasOwnership need a higher level of abstraction to minimize various miscellaneous changes to source code in many places like this if these any rules to hasRead, hasWrite, etc. change (which always happens). From the repeated code block, higher abstraction methods are needed to hide these details: e.g. checkHasEditorAccessType(), checkHasViewerAccessType, checkHasOwnerAccessRight. Again, these authorization checks should be responsibilty of a separate class.

Reply:
Alternatively, you can create just one method called checkPermissionByAccessType(AccessType accessType, Permission permission). Note that a new class called Permission is introduced here to wrap the details of hasRead, hasOwner, hasXXX. It makes the code much cleaner. Btw, as you're adding code relating to these permissions, I suggest we need a little refactoring here: adding the Permission class and replacing all references to details of hasXXX to this Permission class.